src/org/sosy_lab/java_smt/solvers/princess/PrincessAbstractProver.java

```diff
@@ -46,6 +47,15 @@ abstract class PrincessAbstractProver<E> extends AbstractProverWithAllSat<E> {
   protected final PrincessFormulaManager mgr;
   protected final Deque<Level> trackingStack = new ArrayDeque<>(); // symbols on all levels
 
+  /**
+   * Values returned by {@link Model#evaluate(Formula)}.
+   *
+   * <p>We need to record these to make sure that the values returned by the evaluator are
+   * consistant. Calling {@link #isUnsat()} will reset this list as the underlying model has been
+   * updated.
+   */
+  protected final List<IFormula> evaluatedTerms = new ArrayList<>();
+
```

Comment:
This might be a bigger issue for Princess. 
Can we add a plain test for Princess (without this code) to see where Princess fails to be consistant?
Can we ask the authors to provide a consistant model?

Reply:
> This might be a bigger issue for Princess. Can we add a plain test for Princess (without this code) to see where Princess fails to be consistant? Can we ask the authors to provide a consistant model?

I'm afraid in this case the problem here is really with the work-around, and not with Princess. We effectively emulate `(get-values (<expr>))` with a sequence of commands:
```
(push 1)
(assert (= <new_variable> <expr>))
(check-sat)
(get-model) 
// Now get the value from the entry for <new_variable>
(pop 1)
```
The problem is than the sat check causes the solver to forget the old model, and because of this it can choose a different assignment for the variables. For instance, if the only assertion on the stack is `x < y` then calling `evaluate(x)` twice may return different values as the solver simply chooses a different model each time. You can see this by replacing  the code in `PrincessAbstractProver.getEvaluatedTerms()` with `ImmutableList.of()`. This should cause tests in `ModelTest` and `SolverAllSatTest` to fail as the model is not kept consistent across multiply calls to `evaluate()`.

Reply:
The implementation of `PrincessAbstractProver.isUnsat()` should not depend on earlier model evaluations, because it can produce a completely different model. The current implementation seems to work well, even if its runtime is not optimal. We can improve this in the future.